### ISSUE TITLE
Getting hup events right after accept()

### ISSUE THREAD

**Issue by syml** [carllerche/mio#485]
I wrote a multithreaded asynchronous http server in rust using mio. When I run a load test (using siege) the server works fine on the first load test but when the load test is done all requests to the server start failing. Using some logging I noticed that every new connection I get with accept() receives a hangup event immediately. The server connection itself doesn't get any error or hangup events.

I am running rust 1.12.0, mio 0.6 on OS X el capitan

Load test I am using: siege -b -c10 -d10 -t20S http://localhost:8080

Here's the main event loop of my server (full code with example https://github.com/syml/rust-webserver):

``` rust
pub fn run(self) {
    let poll = Poll::new().unwrap();
    let server = TcpListener::bind(&SocketAddr::from_str(&self.host).unwrap()).unwrap();
    poll.register(&server, SERVER, Ready::readable(), PollOpt::edge()).unwrap();
    let mut events = Events::with_capacity(1024);
    let mut next_conn: usize = 1;
    let mut workers = Vec::new();
    // Create worker threads.
    for _ in 0..self.num_workers {
        let (tx, rx) = channel();
        let worker_handler = self.event_handler.duplicate();
        thread::spawn(move || {
            Self::process_events(rx, worker_handler);
        });
        workers.push(tx);
    }
    loop {
        println!("Polling...");
        match poll.poll(&mut events, None) {
            Err(e) => panic!("Error during poll(): {}", e),
            Ok(_) => {}
        }
        for event in events.iter() {
            match event.token() {
                SERVER => {
                    println!("Accepting..");
                    match server.accept() {
                        Ok((stream, _)) => {
                            println!("Registering new connection...");
                            match poll.register(&stream,
                                                Token(next_conn),
                                                Ready::readable(),
                                                PollOpt::edge()) {
                                Err(e) => panic!("Error during register(): {}", e),
                                Ok(_) => {
                                    println!("New connection on worker {} ",
                                             next_conn % self.num_workers);
                                    workers[next_conn % self.num_workers]
                                        .send(Msg::NewConn(next_conn, stream))
                                        .unwrap();
                                    next_conn += 1;
                                }
                            }
                        }
                        Err(e) => panic!("Error during accept() : {}", e),
                    }
                }
                Token(id) => {
                    println!("Sending event on conn {} to worker {}",
                             id,
                             id % self.num_workers);
                    workers[id % self.num_workers]
                        .send(Msg::ConnEvent(id, event.kind()))
                        .unwrap();
                }
            }
        }
    }
}

fn process_events(channel: Receiver<Msg>, mut event_handler: Box<EventHandler>) {
    loop {
        let msg = channel.recv().unwrap();
        match msg {
            Msg::NewConn(id, conn) => {
                event_handler.new_conn(id, conn);
            }
            Msg::ConnEvent(id, event) => {
                event_handler.conn_event(id, event);
            }
        }
    }
}
```

**Comment by matwork** [comment]
I run into the same issue with a similar test case:
- do a stress test involving a lot of connection/disconnections, r/w
- stop it (bulk of HUP events)
- then new connections are getting a HUP event

Another piece of information is that the address then returned by:
`match server.accept() {
                        Ok((stream, addr)) => `
is wrong, it's an address associated with the previous stress load.

**Comment by matwork** [comment]
to fix this issue either:
- set the server socket on level trigger
- loop on server.accept
otherwise incoming connections are handled only once per socket state change.

**Comment by carllerche** [comment]
Thanks @matwork for replying to this. Since it doesn't seem like a bug, I'm going to close the issue. Feel free to open a new issue if you think that there is actually a bug in Mio.
